### ISSUE TITLE
Don't throw exception in ServiceDescriptor

### ISSUE THREAD

**Issue by benjaminpetit** [dotnet/runtime#99622]
Right now, when a `ServiceDescriptor` is not keyed, accessing all the keyed properties (`KeyedImplementationInstance`, `KeyedImplementationType`, `KeyedImplementationFactory`) will throw an `InvalidOperationException`.

Similary, when the `ServiceDescriptor is not keyed, accessing all the non-keyed properties(`ImplementationInstance`, `ImplementationType`, `ImplementationFactory`) will also throw an `InvalidOperationException`.

Ideally the application should check the value of `IsKeyedService` before accessing these properties, but applications that are not yet aware of keyed services break because of these exceptions.

We should remove this behavior. Note that applications that do not "keyed services aware" can still break because all the expected properties will be `null`, but it should cause less issues.

@halter73 @steveharter @davidfowl

**Comment by halter73** [comment]
What's motivating this? A clear exception when accessing the property seems better than a NullReferenceException later in the code. It's a way to ensure people using containers without keyed service get clear error messages.

**Comment by benjaminpetit** [comment]
Some app/dependenciy might scan the list of `ServiceDescriptor` for different reasons.

Check this issue for example: https://github.com/microsoft/ApplicationInsights-dotnet/issues/2852

**Comment by ReubenBond** [comment]
I also think that null is better behavior than an exception. We had 3 nullable one-of properties previously. The best behavior now would be to have 6 (3 keyed, 3 non-keyed) in my opinion. Explosive properties are cumbersome to deal with.

Eg, the most common pattern which is triggering this bad behavior (I already fixed a similar issue in Microsoft.Identity.Web and the same issue existed in Dapr) is to have a marker instance in DI to signal that services have already been added. This common pattern blows up with the new .NET 9 behavior.
